common/web/gesture-recognizer/src/touchEventEngine.ts

```diff
@@ -68,37 +68,76 @@ namespace com.keyman.osk {
-
-      if(!ZoneBoundaryChecker.inputStartOutOfBoundsCheck(coord, this.config)) {
-        // If we started very close to a safe zone border, remember which one(s).
-        // This is important for input-sequence cancellation check logic.
-        this.disabledSafeBounds = ZoneBoundaryChecker.inputStartSafeBoundProximityCheck(coord, this.config);
-      } else {
-        this.disabledSafeBounds = 0;
-        // TODO:  disable tracking for the specific `Touch` that just started.
-        // Requires explicit implementation of multi-touch tracking, which has not yet been added.
+      // If it's not an event we'd consider handling, do not prevent event
+      // propagation!  Just don't process it.
+      if(!this.config.targetRoot.contains(event.target as Node)) {
+        return;
       }
 
-      this.onInputStart(coord);
+      this.preventPropagation(event);
+
+      for(let i=0; i < event.changedTouches.length; i++) {
+        const touch = event.changedTouches.item(i);
+        const sample = this.buildSampleFromTouch(touch);
+
+        if(!ZoneBoundaryChecker.inputStartOutOfBoundsCheck(sample, this.config)) {
+          // If we started very close to a safe zone border, remember which one(s).
+          // This is important for input-sequence cancellation check logic.
+          this.safeBoundMaskMap[touch.identifier] = ZoneBoundaryChecker.inputStartSafeBoundProximityCheck(sample, this.config);
+        } else {
+          // This touchpoint shouldn't be considered; do not signal a touchstart for it.
+          continue;
+        }
+
+        this.onInputStart(touch.identifier, sample, event.target);
+      }
     }
 
     onTouchMove(event: TouchEvent) {
-      this.preventPropagation(event);
-      const coord = InputEventCoordinate.fromEvent(event);
+      for(let i=0; i < event.changedTouches.length; i++) {
+        const touch = event.changedTouches.item(i);
+
+        if(!this.hasActiveSequence(touch.identifier)) {
+          continue;
+        }
+
+        this.preventPropagation(event);
 
-      if(!ZoneBoundaryChecker.inputMoveCancellationCheck(coord, this.config, this.disabledSafeBounds)) {
-        this.onInputMove(coord);
-      } else {
-        this.onInputMoveCancel(coord);
+        const sample = this.buildSampleFromTouch(touch);
+
+        if(!ZoneBoundaryChecker.inputMoveCancellationCheck(sample, this.config, this.safeBoundMaskMap[touch.identifier])) {
+          this.onInputMove(touch.identifier, sample);
+        } else {
+          this.onInputMoveCancel(touch.identifier);
+        }
       }
     }
 
     onTouchEnd(event: TouchEvent) {
-      this.preventPropagation(event);
-      this.onInputEnd(InputEventCoordinate.fromEvent(event));
+      for(let i=0; i < event.changedTouches.length; i++) {
+        const touch = event.changedTouches.item(i);
+
+        if(!this.hasActiveSequence(touch.identifier)) {
+          continue;
+        }
+
+        this.preventPropagation(event);
```

Comment:
This will run multiple times if there is more than 1 changed touch?

Reply:
`TouchEvent.changedTouches` includes _all_ changed touches.  It's a `TouchList`, not a single `Touch`.

Reply:
Ah, I see.  It may; I don't believe this to be causing any problems, though.  I haven't seen JS throw errors at me for calling it more than once, though maybe I should look more closely, just to be sure.  (Note that the multitouch unit test in a descendant PR succeeds without a hitch; we'd expect it to break with an error if this were an issue.)